lib/pbench/agent/tool_meister.py

```diff
@@ -285,12 +275,14 @@ def install(self):
         )
         return (cp.returncode, cp.stdout.strip())
 
-    def start(self):
+    def start(self, tool_dir: Path):
         """Creates the background process running the tool's "start" operation.
         """
-        assert self.tool_dir is not None, "Logic bomb!  no tool directory provided!"
-        if not self.tool_dir.is_dir():
-            raise RuntimeError(f"tool directory does not exist: {self.tool_dir}")
+        assert tool_dir is not None, "Logic bomb!  no tool directory provided!"
```

Comment:
Since this is now an external method parameter, an exception would be much more appropriate than an assert.

An assert is best reserved for "we did something wrong" (and can't recover), not "you did something wrong" since it's going to stop everything in its tracks.

Reply:
We are the we.  So why wouldn't we use an assert to say it was coded incorrectly still?

For that matter, all of these checks are really we did something wrong.  So they really should be asserts.

Reply:
I think of most of Tool Meister as a class library we can (and to some degree probably should) expose for external use; adding asserts makes that a lot more complicated to use.

Reply:
That is a possibility for us to position these classes, and when we do we can setup the interfaces properly.  But I'd like to resist making changes prematurely towards that.

Reply:
One final appeal:

I really dislike seeing `assert` in "inside" code. I think Python will terminate with a 1 on an assert failure, but that undoes all the effort you've taken to categorize errors with specific codes; we won't be handling an exception and reporting some sensible error, we just blow everything up.

Furthermore since this is internal tool meister code running on the remote, (where we actually start tools), it's not clear we'll even see it happen, and it'll be a mess that's a pain to debug. We *should* be sending a failure message up to the TDS to shut everything down gracefully, not just dying.

Reply:
Further, what about https://docs.python.org/3/reference/simple_stmts.html#the-assert-statement (use of `-o`)?

Reply:
In the past I have used "assertions" to document the assumptions a function or method is operating under.  Is that not valid here?

Reply:
I think it is reasonable to use `assert` statements to declare the assumptions and invariants which are part of a function's interface.  However, they come with several liabilities.

1) They are a little bit bulky
  - they take up space on the screen
  - they potentially take up time on the CPU

as such they would be better replaced by in-line code comments and out-of-line unit tests.

2) They look like they do something
 - they look like executable statements, so it becomes tempting to (or at least easy to fall into the trap and) put code inside them which has important side-effects
 - it is tempting/easy to become dependent upon them to catch error conditions or otherwise to protect subsequent code

but, as Peter noted, if the code is invoked with optimization enabled, then no code is produced for `assert` statements, which means that the side-effects and protection are lost!

3) They are likely to interact with the rest of our error handling in unexpected ways.  An assertion failure _does **not**_ immediately kill the program!  It simply raises an `AssertionError` exception.  And, we have a nasty habit of using `catch Exception` clauses in our code, which will _catch that_!  So, instead of crashing the program, an assertion failure will likely be treated just like some other error, and who knows what the result will be....

So, I recommend against using `assert`'s in production code.  I think we should _document_ the function's requirements in the docstring, and test them using unit tests.  (With proper unit testing, any calls to a given function should be replaced by mocks which can assert the preconditions and invariants, so the assertions don't have to be in the production code itself.)

Reply:
I think we should stop catching `Exception` as a general fallback.  Then use assert statements in the code, and have unit tests that verify in testing that when an invariant is broken an assertion fires, and then assertions become useful again.

Reply:
"Assertions" it code comments get stale too quickly.

Reply:
One more attempt. You're saying that

1. There's no need to test for any of these conditions in production deployment. (E.g., if we were deploying with `-O`, which we likely should.)
2. Should any of these exceptions occur in our normal development, it's fine to terminate whatever Pbench command is running with shell status 1. (Which may actually be true, since most scripts will use `!= 0` tests, but it seems to me that this undoes much of your diligence around distinct error codes.)

(And, yes, operationally we have a lot of `except Exception` clauses we'd need to fix, at fairly significant cost and complexity, to avoid the risk or breaking other logic when an assert fires.)

I've always looked as assertions as safeguards and "compiled documentation" for internal invariants; conditions that "can't" happen and signal unrecoverably broken internal state.

I don't believe that applies any time you're testing values sent into a subsystem from outside (including ideally a class method being called by other product code, because that fuzzes modularity and limits unit testing). You clearly do, so I'll stop arguing about it... but I disagree. 🍤

Reply:
> I think we should stop catching `Exception` as a general fallback.

I concur (with a few "exceptions" :wink:), but I think that's largely orthogonal to the issue at hand.

> Then use `assert` statements in the code, and have unit tests that verify in testing that when an invariant is broken an assertion fires, and then assertions become useful again.

I disagree with this.  😆  Assertions _are_ test code...having unit tests which test that the assertions work is like having test code for test code!  🤣 

Putting assertions into production code is d*mn close to asking about the noise made by a tree falling the woods when there is no one around.  Assertions in production code should **_never_** fail...so there is basically no point in having them (which is why `-o` removes them...).  (And, if we're going to have them, then when they fail they should cause the usermode equivalent of a kernel panic -- they shouldn't just raise an exception!)

If we're trying to guard against bad inputs, then we should use explicit checks for that which reliably report the error (this is _not_ what assertions are for).  If we're trying to check for violation of invariants, that's a testing operation, and it should be performed by unit testing or by inserted code during functional testing, not via `assert` statements in production code.

Ideally, we could have assertions in our production code, as a sort of "active documentation", as Dave said.  But, as I said earlier, it requires discipline to avoid the pitfalls associated with that.  And, in order to use the Python-native `assert` statements, we need to ensure that our code doesn't inadvertently catch any of the `AssertionError` exceptions.

So, at least given the current state of our code, I think it would be best to avoid adding `assert` statements to our production code.

Reply:
Can we continue discussing this to find out how we can all be satisfied by a future code change to make this all work?

It'd be great to arrive at an understanding for the team on when the `assert` built-in offered by python is to be used and when not to use it.  It seems really important for us to document and clearly define what we want for the code.

Below is a very long reply trying to address a number of comments made.  It may not be worth to review it, since much of it continues the philosophical discussion and the investigation into the behaviors surrounding this code has led to the finding that the `v0.69` start/stop tools behaviors DO NOT lead to the running bench-script being stopped.  This means that we need to make the start/stop/wait sequence record behaviors and errors and continue the execution cycle.

----

> One final appeal:
> 
> I really dislike seeing `assert` in "inside" code. I think Python will terminate with a 1 on an assert failure, but that undoes all the effort you've taken to categorize errors with specific codes; we won't be handling an exception and reporting some sensible error, we just blow everything up.

Can we find where this behavior of Python 3 is documented?  From [Python 3 Exception Hierarchy](https://docs.python.org/3/library/exceptions.html#exception-hierarchy), an `AssertionError` is a sub-class of `Exception`.  So catching `Exception` will make properly catch and record assertions that fire.
 
> 
> Furthermore since this is internal tool meister code running on the remote, (where we actually start tools), it's not clear we'll even see it happen, and it'll be a mess that's a pain to debug. We _should_ be sending a failure message up to the TDS to shut everything down gracefully, not just dying.

All uncaught exceptions will be ultimately caught at: https://github.com/distributed-system-analysis/pbench/blob/343aa50ef0c99bf1866a21bdea7f5279b2dddf85/lib/pbench/agent/tool_meister.py#L1716  Whether or not the exception is an `AssertionError` or some other exception, the Tool Data Sink will have a problem with such an unexpected exception.

We should probably have a way for the Tool Meister to make some attempt to tell the Tool Data Sink about its problem, but there is no guarantee that it can.  The Tool Data Sink needs to be able to deal with a Tool Meister that disappears regardless of this situation.  The best we can do is what we are doing today: send what logs we can back to the Tool Data Sink directly, write logs locally so that post-mortem debugging can occur.

At best, sending a graceful error message is nice but can't be relied upon.  It is a distributed system.

> One more attempt. You're saying that
> 
> 1. There's no need to test for any of these conditions in production deployment. (E.g., if we were deploying with `-O`, which we likely should.)

Yes.  The code that consumes the `Tool` class hierarchy is in this module, and the `assert` statements should accurately record the assumptions the code is making.

If an `AssertionError` fires without `-O` being used, or some other unexpected error fires with `-O` being used, the full exception traceback is logged, and the same error handling behavior of a lost connection from the Tool Meister occurs.

Why do we need to perform multiple tests that all is order before taking the action in production?  What does it gain us other than adding more checks which slow the code down?

In either case, the error that results still needs to be handled.

> 2. Should any of these exceptions occur in our normal development, it's fine to terminate whatever Pbench command is running with shell status 1. (Which may actually be true, since most scripts will use `!= 0` tests, but it seems to me that this undoes much of your diligence around distinct error codes.)

As above, if any assertion fires during normal development, the assertion message should help us understand what invariant is violated.

The return code for the Tool Meister code will be `10`: https://github.com/distributed-system-analysis/pbench/blob/343aa50ef0c99bf1866a21bdea7f5279b2dddf85/lib/pbench/agent/tool_meister.py#L1718

The diligence about distinct errors codes originated from debugging Tool Meister problems in containers.

In containerized environments, if the containerized environment is not constructed carefully, log output or stdout/stderr can easily be lost, but the container run times often still capture the process exit code.  This was a tremendous help determining what happened.

> 
> (And, yes, operationally we have a lot of `except Exception` clauses we'd need to fix, at fairly significant cost and complexity, to avoid the risk or breaking other logic when an assert fires.)

Not sure what this means but as above, `AssertionError` is a sub-class of `Exception`.

> 
> I've always looked as assertions as safeguards and "compiled documentation" for internal invariants; conditions that "can't" happen and signal unrecoverably broken internal state.

Seems like we are saying something similar.

The code here is using an assert to mean, "this code is assuming that this invariant is held; if violated with assertions enabled we'll tell you about it, if violated with assertions disabled, some other unexpected error will occur."  And the cost of all those assertions is not worth making production code as the unexpected errors will need to be handled anyways.

> 
> I don't believe that applies any time you're testing values sent into a subsystem from outside (including ideally a class method being called by other product code, because that fuzzes modularity and limits unit testing).

Not sure it is so cut and dry.  Is it really worth testing and raising a specific exception every time we write an interface?  In most cases, an error is an error: something didn't work and the error is raised.  The error handling still has to be there.

For all the assertions in this module, what happens if they are just removed entirely?  Will the code stop working?  Will the error handling work differently?  If they are converted into explicit tests with an Exception raised, are all those tests really worth being made each time those methods are called?

The assertions seem to be a good way to: 1. document the invariants, 2. get the tests out of the way for production code.

> So, I'm finally reviewing the code (rather than just joining a philosophical discussion), and I concur with Dave: the code in question is checking the input parameter to this function; I think that that would best be done with a conventional bit of code which reports any error to the caller, rather than by an `assert` -- if there is something wrong with it, I think the initial fault lies with the caller, not with the current function.

The `base-tool` already checks the `--dir` parameter constructed with the string value of the `tool_dir` parameter. Starting and stopping tools today (in 0.69 code base) does not result in an error.  It would be a significant change to have start/stop from Tool Meister do something different without giving the user an option to enable / disable a behavior.

So we really should change the Tool Meister code to not do any checks and just record each step.

Reply:
Which we apparently already do.

Reply:
> Can we find where this behavior of Python 3 is documented? From [Python 3 Exception Hierarchy](https://docs.python.org/3/library/exceptions.html#exception-hierarchy), an `AssertionError` is a sub-class of `Exception`. So catching `Exception` will make properly catch and record assertions that fire.

I'm increasingly confused about the parameters of this discussion, I'm afraid.

When Webb pointed out that we use `except Exception` which would indeed catch your `AssertionError`, you went off and started breaking up the `except Exception` into (slightly) more specific Exception sub-classes, excluding `AssertionError`. I interpreted this as a step towards making `assert` actually terminate.

However, apparently your plan is to use `assert` as just a syntactic `raise AssertionError` which you *expect* to be caught and finalized within our code instead of terminating the program. (Granted, with the extra semantic that, if we ever build with `-O` it'll become just documentation.) Since the official Python documentation states pretty clearly that `assert expression` is the same as `if not expression: raise AssertionError` (and yes, it claims that it raises the class, not an instance, although that appears to be incorrect as a simple test shows the repr as `AssertionError()`)

I think maybe I like the idea of planning to finalize `AssertionError` even less, but I'm getting a headache and I'm ready to stop thinking about this to the point of letting it go without further discussion if you're really determined.

Reply:
In my opinion:
 - "External interfaces" should check their parameters and report errors in them to their callers; there are exceptions to that principle (e.g., when part of the interface contract includes the caller accepting responsibility for failure due to bad inputs), but I would counsel against that approach in all but the most performance-critical cases because of the various problems that it produces when things go bad.  Nevertheless, for this sort of checking using an `assert` would be a totally inappropriate approach:  raising `AssertionError` should not be part of the interface contract in any case.
- Invariant checks shouldn't be present in production code unless there is a real chance of them failing or unless they are needed to protect a critical resource.

So, for the code in question here, `tool_dir` is a parameter provided by the caller, and this is not a performance-critical interface, so it should be checked unless the interface is not "external".  (I don't know whether we want to make the "internal" vs. "external" determination here/now....)  A similar logic applies to `self`:  although it is an implicit parameter, it is nevertheless still the case that this function is a method invoked on that object, this instance of the object is provided by the caller, and, given that this is Python, there's no reason to believe anything about the object until we check it, if we deem this to be an "external" interface.

At the same time, the states of `self.tool_dir`, `self.start_process`, and `self.stop_process` represent internal invariants (assuming that the object is valid); however, I think that, through testing, we can establish that there is no real chance of their being violated.  And, since failing to maintaining them does not, as far as I can see, put any critical infrastructure or financial position at risk, I don't think the live, production code needs to check them.

All that said, I'm can make peace with having various `assert` statements here (after all, they should never fail, and if we run with `-O`, they won't even be present).  But, until we have the rest of the exception handling stuff in order, I don't think it's a good idea, because I don't think they are better than code comments, and, if one does fail, it's not clear that it will make the situation any easier to understand.  And, I don't think I want to see this become common practice...I think we should be relying on unit-testing for that.